staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go

```diff
@@ -974,9 +995,9 @@ func (alloc *allocator) allocateDevice(r deviceIndices, device deviceWithID, mus
 
 	// The API validation logic has checked the ConsumesCounters referred should exist inside SharedCounters.
 	if alloc.features.PartitionableDevices && len(device.basic.ConsumesCounters) > 0 {
-		// If a device consumes capacity from a capacity pool, verify that
```

Comment:
I think the existing logic around feature gate checking is sub-optimal here.

If a device consumes counters and the feature is disabled, we should skip the device and continue looking for a device which does not need the feature.

What the code does right now is ignore the field and use the device. Nothing good will come out of that...

Reply:
Yeah, this doesn't look right. But we actually filter out all ResourceSlices with shared counters when we determine the eligible pools and validation doesn't allow `ConsumesCounters` entries that doesn't reference a valid counter set.
But it was confusing, so I've fixed the logic here and also added comment that it is a situation that shouldn't happen. We also already have a test for this scenario.